File: indhu/leetcode/1_string/4_771.js
// https://leetcode.com/problems/jewels-and-stones/description/

class Solution {

    fn(jewels, stones) {
        let result = 0;
        for (let i = 0; i < stones.length; i++) {
            if (jewels.includes(stones[i])) {
                result++
            }
        }
        return result
    }

}

const data = new Solution().fn("aA", "aAAbbbb");
console.log(data);

// Observation:
// input: jewels unique, case sensitive
// output: jewels contains in stones
// topic: string, hash table(hold)
// Methods: includes [O(n)]
// output Time Complexity : O(n^2)


// Alternate
// Set.has() [O(1)]

File: usha/leetcode/4_HashTable/easy/4_771.java
// https://leetcode.com/problems/jewels-and-stones/description/

// Topics
// Hash Table
// String


class Solution {
    public int numJewelsInStones(String jewels, String stones) {
        boolean[] isJewel = new boolean[128]; // ASCII size

    for (char jewel : jewels.toCharArray()) {
        isJewel[jewel] = true;
    }

    int count = 0;
    for (char stone : stones.toCharArray()) {
        if (isJewel[stone]) {
            count++;
        }
    }
    return count;
    //     int count = 0; // approach I tried - 2ms- acceptance- 47.79
    // for (char stone : stones.toCharArray()) {
    //     if (jewels.indexOf(stone) != -1) {
    //         count++;
    //     }
    // }
    // return count;
    }
}